### PR TITLE
Make secp256k1 optional

### PR DESCRIPTION
**Pull request by tcharding**
Add feature "crypto" and feature gate anything that touches `secp256k1` behind it.

Fix: #526 

### Note

Does not add `doc(cfg` for private functions, do we want that?

### CODE REVIEW

**Comment by Kixunil** [comment]
Cool! Needs `doc(cfg(...))` in public items though.

**Comment by tcharding** [comment]
Doing this I learned that `doc(cfg` has no effect on enum variants.

**Comment by tcharding** [comment]
Observation: The feature guarded imports are pretty ugly.

**Comment by Kixunil** [comment]
> Doing this I learned that `doc(cfg` has no effect on enum variants.

WTF, that sucks! We will need to document it in text then.

> The feature guarded imports are pretty ugly.

You can decrease the ugliness by putting guarded impls into separate modules:

```rust
pub struct Foo {
    // ...
}

impl Foo {
    // unconditional impls
}

#[cfg(feature = "secp256k1")]
#[cfg_attr(docsrs, doc(cfg = "secp256k1"))] // please check that this doc is propagated to the items inside, I'm not sure it is
mod secp256k1_impls {
    use secp256k1::PublicKey;

    // Yes, this works fine, the restriction is that it has to be the same crate
    impl Foo {
        // Something that uses Public key here
    }
}
```

**Comment by tcharding** [comment]
> Went to report it and when I wrote a minimal reproducer I saw it documented correctly

Face palm, no idea where I looked yesterday - it is as you say when I hacked up an example just now. My bad.

> You can decrease the ugliness by putting guarded impls into separate modules:

That's a good idea.

**Comment by tcharding** [comment]
Added docs attributes to enums. I went over the diff and didn't see any chunks of code that would be better in sub-modules.

**Comment by tcharding** [comment]
Rebase only on other changes.

**Comment by tcharding** [comment]
CI was red so I rebased and checked/re-did a bunch of the feature guards on import statements.

File: bitcoin/Cargo.toml [hunk 1]
```diff
@@ -14,11 +14,13 @@ edition = "2018"
 
 # Please don't forget to add relevant features to docs.rs below
 [features]
-default = [ "std", "secp-recovery" ]
-rand = ["secp256k1/rand-std"]
-serde = ["actual-serde", "bitcoin_hashes/serde", "secp256k1/serde"]
+default = ["std"]
+crypto-std = ["secp256k1/std", "crypto"] # Enable this feature to use ECC (i.e. keys and sigs).
+crypto = ["secp256k1/alloc"]
```
Comment (Kixunil) [inline bitcoin/Cargo.toml:19]: We should depend on `alloc` in `secp256k1` specification below and call the feature `secp256k1`. We only need `std` for `Error`, right?

**Review by Kixunil — CHANGES_REQUESTED** [review]
Can't review properly but at least I found non-idiomatic naming.

File: bitcoin/Cargo.toml (hunk 1 above)
Comment (tcharding) [reply bitcoin/Cargo.toml:19]: I called it `crypto` thinking that we were about move the `crypto` module to a crate and then this feature would enable stuff from that crate and from `secp25k1`, does that sound reasonable?

Comment (Kixunil) [reply bitcoin/Cargo.toml:19]: Oh, OK then, if the crypto crate PR gets in the feature name should stay.

**Comment by Kixunil** [comment]
@TheBlueMatt compile time of whole `secp256k1` is quite significant since it also compiles the build script with its dependencies and requires C compiler which may be annoying to get to work on unusual platforms. Further short compile times are interesting for package vendors since new versions of software are generally compiled from scratch.

**Comment by Kixunil** [comment]
I package a few Debian packages that do use `rust-bitcoin` and the trouble is that sometimes if something goes wrong I have to recompile everything. `electrs` uses LTO for instance. So decreasing the compile time is quite interesting. But if multiple people are against it I would respect that.

**Comment by tcharding** [comment]
I have zero opinions on this, will wait for others to weigh in before rebasing or closing.

**Comment by tcharding** [comment]
Since I do not have strong feelings and no one has weighed in with any, leaving for another day.

**Comment by Kixunil** [comment]
I don't want to push this overly strongly but I just remembered one more argument: dealing with C crap is often painful experience, especially when cross-compiling or on less common OSes. Being able to avoid it can help. We had a bunch of issues at electrs, although admittedly, they were caused by `rocksdb`, not `secp256k1`. And yes, `secp256k1` is quite painless overall, but still there's some risk.

**Comment by apoelstra** [comment]
Let's revisit once the crate smashing is done. My feeling with this is "only if it's easy". I agree that secp256k1 is likely to be a particularly big pain point, but OTOH it's also responsible for the bulk of this library's functionality. For example, after crate smashing we may have a `bitcoin-crypto` crate that we could simply make optional and be done with it.

If we try to do this now it's just going to be another complication as we're rearranging other stuff.

**Comment by Kixunil** [comment]
OK, that makes sense.

**Comment by RCasatta** [comment]
I am trying to build a statically linked executable that has rust-bitcoin as dep with musl. It doesn't need secp256k1 and it is failing on building it so it may be one reason to have this.

The other option is understanding why the build fails with a strange error `  error occurred: Failed to find tool. Is `aarch64-linux-musl-gcc` installed?` while I am trying to target  `aarch64-unknown-linux-musl` 

<details>
<summary>Details</summary>
<code>
error: failed to run custom build command for `secp256k1-sys v0.8.1`

Caused by:
  process didn't exit successfully: `/home/casatta/git/dinasty/target/release/build/secp256k1-sys-7ab9d397732e17b3/build-script-build` (exit status: 1)
  --- stdout
  TARGET = Some("aarch64-unknown-linux-musl")
  OPT_LEVEL = Some("3")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_aarch64-unknown-linux-musl
  CC_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CC_aarch64_unknown_linux_musl
  CC_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64-unknown-linux-musl
  CFLAGS_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64_unknown_linux_musl
  CFLAGS_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = Some("neon")
  cargo:rerun-if-env-changed=CC_aarch64-unknown-linux-musl
  CC_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CC_aarch64_unknown_linux_musl
  CC_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64-unknown-linux-musl
  CFLAGS_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64_unknown_linux_musl
  CFLAGS_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("neon")
  running: "aarch64-linux-musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-I" "depend/secp256k1/" "-I" "depend/secp256k1/include" "-I" "depend/secp256k1/src" "-Wall" "-Wextra" "-DSECP256K1_API=" "-DENABLE_MODULE_ECDH=1" "-DENABLE_MODULE_SCHNORRSIG=1" "-DENABLE_MODULE_EXTRAKEYS=1" "-DECMULT_GEN_PREC_BITS=4" "-DECMULT_WINDOW_SIZE=15" "-DUSE_EXTERNAL_DEFAULT_CALLBACKS=1" "-DENABLE_MODULE_RECOVERY=1" "-o" "/home/casatta/git/dinasty/target/aarch64-unknown-linux-musl/release/build/secp256k1-sys-ee85cd798ceda1c6/out/depend/secp256k1/contrib/lax_der_parsing.o" "-c" "depend/secp256k1/contrib/lax_der_parsing.c"
  cargo:rerun-if-env-changed=CC_aarch64-unknown-linux-musl
  CC_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CC_aarch64_unknown_linux_musl
  CC_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64-unknown-linux-musl
  CFLAGS_aarch64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64_unknown_linux_musl
  CFLAGS_aarch64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("neon")
  running: "aarch64-linux-musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-I" "depend/secp256k1/" "-I" "depend/secp256k1/include" "-I" "depend/secp256k1/src" "-I" "wasm/wasm-sysroot" "-Wall" "-Wextra" "-DSECP256K1_API=" "-DENABLE_MODULE_ECDH=1" "-DENABLE_MODULE_SCHNORRSIG=1" "-DENABLE_MODULE_EXTRAKEYS=1" "-DECMULT_GEN_PREC_BITS=4" "-DECMULT_WINDOW_SIZE=15" "-DUSE_EXTERNAL_DEFAULT_CALLBACKS=1" "-DENABLE_MODULE_RECOVERY=1" "-o" "/home/casatta/git/dinasty/target/aarch64-unknown-linux-musl/release/build/secp256k1-sys-ee85cd798ceda1c6/out/depend/secp256k1/contrib/lax_der_parsing.o" "-c" "depend/secp256k1/contrib/lax_der_parsing.c"

  --- stderr


  error occurred: Failed to find tool. Is `aarch64-linux-musl-gcc` installed?


warning: build failed, waiting for other jobs to finish...

</code>
</details>

**Comment by Kixunil** [comment]
@RCasatta interesting. I think the message comes from `cc` and that crate is most likely the one to file an issue against.

**Pull request closed without merging**